extensions/sql/transfer-process-store/sql_tps_proposal.md

```diff
@@ -0,0 +1,133 @@
+    data_destination  varchar                                            not null,
+    managed_resources boolean default true,
+    properties        varchar,
+    transfer_type     varchar
+);
+
+comment on column edc.data_request.data_destination is 'DataAddress serialized as JSON';
+
+comment on column edc.data_request.properties is 'java Map serialized as JSON';
+
+comment on column edc.data_request.transfer_type is 'TransferType serialized as JSON';
+
+alter table edc.data_request
+    owner to "user";
+
+create table edc.transfer_process
+(
+    id                       varchar           not null
+        constraint transfer_process_pk
+            primary key,
+    type                     integer default 0 not null,
+    state                    integer           not null,
+    state_count              integer default 0 not null,
+    state_time_stamp         timestamp,
+    trace_context            varchar,
+    error_detail             varchar,
+    resource_manifest        varchar,
+    provisioned_resource_set varchar,
+    data_request_id          varchar           not null
+        constraint data_request___fk
+            references edc.data_request
+            on update restrict on delete restrict
+);
+
+comment on column edc.transfer_process.trace_context is 'Java Map serialized as JSON';
+
+comment on column edc.transfer_process.resource_manifest is 'java ResourceManifest serialized as JSON';
+
+comment on column edc.transfer_process.provisioned_resource_set is 'ProvisionedResourceSet serialized as JSON';
+
+alter table edc.transfer_process
+    owner to "user";
+
+create unique index transfer_process_id_uindex
+    on edc.transfer_process (id);
+
+create unique index data_request_id_uindex
+    on edc.data_request (id);
+
+```
+
+## Translating the `TransferProcessStore` interface into SQL statements
+#### `processIdForTransferId`
+selects a TransferProcess based on the DataRequest ID
+```sql
+select t.* from edc.transfer_process t
+where t.data_request_id=
+      (select d.id
+        from edc.data_request d
+        where d.process_id = 'test-pid');
+```
+
+#### `find`
+selects a TP based on its ID
+```sql
+select t.* 
+from edc.transfer_process t
+where t.id='test-id1';
+```
+
+#### `nextForState`
+Selects the next batch of TP rows that are in a particular state and locks them. 
+```sql
+select * from edc.transfer_process where state <> 600 limit 5 for update skip locked ;
+-- do TPM asynchronous stuff
+update edc.transfer_process t set state = 800 where id='test-id2'
+```
+The `skip locked` statement will prevent any subsequent processes from blocking. It will most closely mimic the behaviour of the `lease`.
+
+_Nota bene: this locks a row until the transaction is completed, see [below](#open-questions) for details._ 
+
+
+#### `create`
+inserts a new TP into the database by first inserting the `DataRequest`, obtaining it's ID and then creating the `TransferProcess` entry.
+```sql
+with drq_id as
```

Comment:
I assume `WITH` is also an extension to the ANSI/ISO SQL standard.

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Daimler TSS GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

Reply:
Thanks for pointing this out, I didn't know that. Maybe our committers can decide which SQL standard or databases the EDC needs to support. I think this was never specified.

I also found this in [Wikipedia](https://en.wikipedia.org/wiki/SQL:1999):

> Recursive CTEs from the standard were relatively close to the existing implementation in [IBM DB2](https://en.wikipedia.org/wiki/IBM_DB2) version 2.[[1]](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#cite_note-JimMelton1-1) Recursive CTEs are also supported by [Microsoft SQL Server](https://en.wikipedia.org/wiki/Microsoft_SQL_Server) (since SQL Server 2008 R2),[[2]](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#cite_note-msdnRCTEs-2) [Firebird 2.1](https://en.wikipedia.org/wiki/Firebird_(database_server)),[[3]](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#cite_note-3) [PostgreSQL 8.4+](https://en.wikipedia.org/wiki/PostgreSQL),[[4]](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#cite_note-4) [SQLite 3.8.3+](https://en.wikipedia.org/wiki/SQLite),[[5]](https://en.wikipedia.org/wiki/Hierarchical_and_recursive_queries_in_SQL#cite_note-5) [IBM Informix](https://en.wikipedia.org/wiki/IBM_Informix) version 11.50+, [CUBRID](https://en.wikipedia.org/wiki/CUBRID), [MariaDB 10.2+](https://en.wikipedia.org/wiki/MariaDB) and [MySQL 8.0.1+](https://en.wikipedia.org/wiki/MySQL).

And it looks like Derby DB doesn't support it ([source](https://issues.apache.org/jira/browse/DERBY-11)).

Reply:
Recursive CTL is something very special - here we have non-recursive CTE

Reply:
Since we're using user-supplied IDs (as opposed to: DB-generated ones), we don't need the `returning id` part. Therefor I don't see a problem in replacing the `WITH` with two consecutive SQL statements as long as they are executed in the same transaction. @algattik do you agree, am I missing something?